In [1]:
import sys
from pathlib import Path
import os
import json

sys.path.append(str(Path.cwd().parent.parent))
print(str(Path.cwd().parent.parent))
print("Current working directory: ", os.getcwd())

from ModelDevelopment.CentralPipeline import CentralPipeline
from ModelDevelopment.ImageBatchPipeline import ImageBatchPipeline
from DataProcessing.DataPreProcessing import DataPaths

c:\Users\colin\OneDrive\Desktop\Jersey-Number-Recognition
Current working directory:  c:\Users\colin\OneDrive\Desktop\Jersey-Number-Recognition\ModelDevelopment\experiments


In [2]:
tracklets_to_process_override_batch1 = [i for i in range(0, 711)] # COLIN
tracklets_to_process_override_batch2 = [i for i in range(712, 1426)] # JARED Second half of the tracklets (605 items)
#merge_cache = [i for i in range(0, 1426)]

pipeline = CentralPipeline(
  num_tracklets=1426,
  #num_images_per_tracklet=50,
  #tracklets_to_process_override=tracklets_to_process_override_batch1,
  input_data_path=DataPaths.CHALLENGE_DATA_DIR.value,
  output_processed_data_path=DataPaths.PROCESSED_DATA_OUTPUT_DIR_CHALLENGE.value,
  common_processed_data_dir=DataPaths.COMMON_PROCESSED_OUTPUT_DATA_CHALLENGE.value,
  gt_data_path="", #DataPaths.TEST_DATA_GT.value,
  use_cache=True, # Set to false to rebuild the cache
  suppress_logging=False,
  use_image_enhancement=False,
  
  # --- PARALLELIZATION PARAMS --- These settings are optimal for an NVIDIA RTX 3070 Ti Laptop GPU.
  num_workers=2,                   # CRITICAL optimisation param. Adjust accordingly. STR: 1 | Everything else: 6
  tracklet_batch_size=32,          # CRITICAL optimisation param. Adjust accordingly. Always 32
  image_batch_size=1000,           # CRITICAL optimisation param. Adjust accordingly.
  num_threads_multiplier=6         # CRITICAL optimisation param. Adjust accordingly.
  )                                # -> Interpretation: num_threads = num_workers * num_threads_multiplier

2025-04-02 20:11:00 [INFO] DataPreProcessing initialized. Universe of available data paths:
2025-04-02 20:11:00 [INFO] ROOT_DATA_DIR: c:\Users\colin\OneDrive\Desktop\Jersey-Number-Recognition\data\SoccerNet\jersey-2023\extracted
2025-04-02 20:11:00 [INFO] TEST_DATA_GT: c:\Users\colin\OneDrive\Desktop\Jersey-Number-Recognition\data\SoccerNet\jersey-2023\extracted\test\test_gt.json
2025-04-02 20:11:00 [INFO] TRAIN_DATA_GT: c:\Users\colin\OneDrive\Desktop\Jersey-Number-Recognition\data\SoccerNet\jersey-2023\extracted\train\train_gt.json
2025-04-02 20:11:00 [INFO] TEST_DATA_DIR: c:\Users\colin\OneDrive\Desktop\Jersey-Number-Recognition\data\SoccerNet\jersey-2023\extracted\test\images
2025-04-02 20:11:00 [INFO] TRAIN_DATA_DIR: c:\Users\colin\OneDrive\Desktop\Jersey-Number-Recognition\data\SoccerNet\jersey-2023\extracted\train\images
2025-04-02 20:11:00 [INFO] CHALLENGE_DATA_DIR: c:\Users\colin\OneDrive\Desktop\Jersey-Number-Recognition\data\SoccerNet\jersey-2023\extracted\challenge\images
2

In [3]:
#pipeline.aggregate_legibility_results_data()
#pipeline.aggregate_pose()
#pipeline.aggregate_str_results()

In [ ]:
pipeline.run_soccernet(
  run_soccer_ball_filter=False,
  generate_features=True,
  run_filter=False,
  run_legible=True,
  run_legible_eval=False,
  run_pose=True,
  run_crops=True,
  run_str=True,
  run_combine=False,
  run_eval=False,
  use_clip4str=True,
  pyscript=False,
  )

2025-04-02 20:11:01 [INFO] Running the SoccerNet pipeline.
2025-04-02 20:11:01 [INFO] No tracklets provided. Retrieving from input folder.
2025-04-02 20:11:01 [INFO] ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
2025-04-02 20:11:01 [INFO] Min tracklet: 0
2025-04-02 20:11:01 [INFO] Max tracklet: 1425


  0%|          | 0/45 [00:00<?, ?it/s]

2025-04-02 20:11:01 [INFO] Tracklet batch size: 32
2025-04-02 20:11:01 [INFO] Image batch size: 1000
2025-04-02 20:11:01 [INFO] Number of workers: 2
2025-04-02 20:11:01 [INFO] Number of threads created: 12
2025-04-02 20:11:01 [INFO] All tracklets in 0-32 are cached. Skipping feature generation.
2025-04-02 20:11:01 [INFO] All tracklets in 32-64 are cached. Skipping feature generation.
2025-04-02 20:11:01 [INFO] All tracklets in 64-96 are cached. Skipping feature generation.
2025-04-02 20:11:01 [INFO] All tracklets in 96-128 are cached. Skipping feature generation.
2025-04-02 20:11:01 [INFO] All tracklets in 128-160 are cached. Skipping feature generation.
2025-04-02 20:11:01 [INFO] All tracklets in 160-192 are cached. Skipping feature generation.
2025-04-02 20:11:01 [INFO] All tracklets in 192-224 are cached. Skipping feature generation.
2025-04-02 20:11:01 [INFO] All tracklets in 224-256 are cached. Skipping feature generation.
2025-04-02 20:11:01 [INFO] All tracklets in 256-288 are ca

Processing tracklets (CUDA + CPU):   0%|          | 0/10 [00:00<?, ?it/s]

2025-04-02 20:11:14 [INFO] Creating placeholder data files for Soccer Ball Filter.


Processing Batch Tracklets (1344-1376):   0%|          | 0/10 [00:00<?, ?it/s]

2025-04-02 20:11:47 [INFO] Processed tracklet: 1367
2025-04-02 20:11:49 [INFO] Processed tracklet: 1366
2025-04-02 20:12:11 [INFO] Processed tracklet: 1368
2025-04-02 20:12:13 [INFO] Processed tracklet: 1369
2025-04-02 20:12:26 [INFO] Processed tracklet: 1370
2025-04-02 20:12:28 [INFO] Processed tracklet: 1371
2025-04-02 20:12:40 [INFO] Processed tracklet: 1373
2025-04-02 20:12:42 [INFO] Processed tracklet: 1372
2025-04-02 20:12:53 [INFO] Processed tracklet: 1375
2025-04-02 20:13:05 [INFO] Processed tracklet: 1374
2025-04-02 20:13:06 [INFO] Skipping first 31 cached tracklets in batch 1376-1408. Processing the rest.
2025-04-02 20:13:06 [INFO] Using double parallelization: multiprocessing + CUDA batch processing.


Processing tracklets (CUDA + CPU):   0%|          | 0/1 [00:00<?, ?it/s]

2025-04-02 20:13:13 [INFO] Creating placeholder data files for Soccer Ball Filter.


Processing Batch Tracklets (1376-1408):   0%|          | 0/1 [00:00<?, ?it/s]

2025-04-02 20:13:41 [INFO] Processed tracklet: 1407
2025-04-02 20:13:42 [INFO] All tracklets in 1408-1426 are cached. Skipping feature generation.
2025-04-02 20:13:42 [INFO] Generating json for pose
2025-04-02 20:13:42 [INFO] Aggregating legible & illegible results (cache not used or only one file is missing).
2025-04-02 20:13:53 [INFO] Saved global legible results to: c:\Users\colin\OneDrive\Desktop\Jersey-Number-Recognition\data\SoccerNet\jersey-2023\processed_data\challenge\common_data\legible_results.json
2025-04-02 20:13:53 [INFO] Saved global illegible results to: c:\Users\colin\OneDrive\Desktop\Jersey-Number-Recognition\data\SoccerNet\jersey-2023\processed_data\challenge\common_data\illegible_results.json
2025-04-02 20:13:53 [INFO] Legible tracklets list: 0, 1, 2, 3, 4, 5, 9, 10, 11, 13, 14, 20, 21, 22, 25, 26, 27, 29, 31, 32, 33, 34, 35, 36, 37, 38, 39, 43, 44, 45, 46, 47, 48, 49, 50, 51, 53, 54, 55, 56, 57, 59, 60, 61, 62, 63, 66, 67, 68, 69, 70, 72, 73, 74, 77, 78, 83, 84, 85

Generating pose JSON:   0%|          | 0/1426 [00:00<?, ?it/s]

2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 8 with 0 images.
2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 19 with 0 images.
2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 15 with 0 images.
2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 17 with 0 images.
2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 0 with 1 images.
2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 18 with 0 images.
2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 16 with 0 images.
2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 12 with 0 images.
2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 6 with 0 images.
2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 7 with 0 images.
2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 20 with 1 images.
2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 14 with 9 images.
2025-04-02 20:13:53 [INFO] Generated pose JSON for tracklet 23 with 

Running pose estimation:   0%|          | 0/1035 [00:00<?, ?it/s]

## Benchmark (current pipeline metrics)
### Results
- 2025-03-29 14:44:05 [INFO] Correct 1051 out of 1210. Accuracy 86.85950413223141%.
- 2025-03-29 14:44:05 [INFO] TP=766, TN=285, FP=70, FN=89
- 2025-03-29 14:44:05 [INFO] Precision=0.916267942583732, Recall=0.895906432748538
- 2025-03-29 14:44:05 [INFO] F1=0.9059727971614429
- 2025-03-29 14:44:05 [INFO] Saved failed legibility cases to:  c:\Users\colin\OneDrive\Desktop\Jersey-Number-Recognition\data\SoccerNet\jersey-2023\processed_data\test\common_data\failed_legibility_cases.json

### Analysis
legible_but_marked_illegible: 89 cases
109, 88, 34, 151, 19, 68, 67, 128, 57, 178, 189, 196, 214, 242, 258, 271, 304, 327, 320, 352, 375, 378, 414, 416, 439, 425, 451, 453, 454, 466, 499, 511, 513, 526, 528, 567, 576, 594, 619, 623, 646, 687, 682, 709, 711, 717, 739, 758, 767, 795, 788, 840, 843, 832, 841, 852, 847, 851, 856, 850, 868, 866, 871, 901, 902, 904, 900, 907, 928, 939, 944, 961, 984, 983, 1005, 1039, 1064, 1063, 1077, 1091, 1101, 1137, 1152, 1153, 1160, 1167, 1157, 1165, 1206

illegible_but_marked_legible: 70 cases
168, 167, 162, 97, 39, 87, 89, 134, 61, 172, 164, 199, 215, 260, 267, 269, 281, 276, 288, 295, 317, 325, 340, 345, 374, 394, 388, 386, 391, 406, 408, 434, 437, 459, 471, 514, 532, 558, 683, 748, 770, 790, 802, 827, 861, 869, 879, 897, 895, 909, 964, 981, 999, 1023, 1031, 1025, 1033, 1030, 1055, 1088, 1087, 1169, 1175, 1174, 1173, 1183, 1170, 1179, 1184, 1185

- Total misclassifications: 159
- False negatives (legible marked as illegible): 89
- False positives (illegible marked as legible): 70
- Percentage of false negatives: 7.36%
- Percentage of false positives: 5.79%
- Combined percentage of errors as a percentage of total cases: 13.14%

In [ ]:
# import json
# failed_cases = f"{DataPaths.COMMON_PROCESSED_OUTPUT_DATA_CHALLENGE.value}/failed_legibility_cases.json"

# # Load this JSON file
# with open(failed_cases, 'r') as f:
#     failed_cases_dict = json.load(f)
#     # Print the every key and its associated value (value will be a list so need to make that printable with ",".join):
#     for key, value in failed_cases_dict.items():
#         print(f"{key}: {len(value)} cases")
#         if len(value) > 0:
#             print(f"{', '.join(value)}")
#         print()
    
#     # Summary statistics
#     legible_but_marked_illegible = len(failed_cases_dict.get("legible_but_marked_illegible", []))
#     illegible_but_marked_legible = len(failed_cases_dict.get("illegible_but_marked_legible", []))
#     total_errors = legible_but_marked_illegible + illegible_but_marked_legible
    
#     print(f"Total misclassifications: {total_errors}")
#     print(f"False negatives (legible marked as illegible): {legible_but_marked_illegible}")
#     print(f"False positives (illegible marked as legible): {illegible_but_marked_legible}")
#     print(f"Percentage of false negatives: {legible_but_marked_illegible / 1210 * 100:.2f}%")
#     print(f"Percentage of false positives: {illegible_but_marked_legible / 1210 * 100:.2f}%")
#     print(f"Combined percentage of errors as a percentage of total cases: {total_errors / 1210 * 100:.2f}%")

In [ ]:
# import timm
# import torch
# from torch import nn

# model = timm.create_model("timm/vit_base_patch16_224.orig_in21k_ft_in1k", pretrained=False)
# model.head = nn.Linear(model.head.in_features, 10)
# model.load_state_dict(
#     torch.hub.load_state_dict_from_url(
#         "https://huggingface.co/edadaltocg/vit_base_patch16_224_in21k_ft_svhn/resolve/main/pytorch_model.bin",
#         map_location="cpu",
#         file_name="vit_base_patch16_224_in21k_ft_svhn.pth",
#     )
# )